### PR TITLE
Jr/APPEALS-39664

### PR DESCRIPTION
**Pull request by Jruuuu**
Resolves [APPEALS-39664](https://jira.devops.va.gov/browse/APPEALS-39664)

# Description
As a Developer, I need to create UpdateVacolsOnOptIn sub service class and logic to update VACOLS and Caseflow DB for SOC/SSOC opt in. Within the Parent Service Class DecisionReviewCreated when we receive a Request Issue that is being SOC/SSOC opt in approved we created a perform! module that updates the VACOLS ISSUE and BREIFF DB to close and Caseflow DB 

## Acceptance Criteria

- [x] Build out UpdatedVacolsOnOptIn Sub Service class that will be in services/events/DecisionReviewCreate/ folder
- [x] This service class will have the namespace of DecisionReviewCreate::UpdatedVacolsOnOptIn , use module to wrap class and namespace
- [x] Implement Logic to check for SOC/SSOC opt in in this sub service class
- [x] Implement logic for updating Vacols and Casflow if SOC/SSOC optins in this new sub serivce class
- [x] Make a call from the main DecisionReviewCreate class to this new subclass
- [x] Process and handle all errors and also manage transactions 
- [x] Create RSPEC scenarios for all above AC

## Testing Plan

1.  [APPEALS-40967-Test Plan](https://jira.devops.va.gov/browse/APPEALS-40967)
2. [APPEALS-40967-X-ray](https://jira.devops.va.gov/secure/XrayExecuteTest!default.jspa?testExecIssueKey=APPEALS-40967&testIssueKey=APPEALS-40967)

# Best practices
## Code Documentation Updates
- [x] Add or update code comments at the top of the class, module, and/or component.

### Test Coverage
Did you include any test coverage for your code? Check below:
- [x] RSpec

### Code Climate
Your code does not add any new code climate offenses? If so why?
- [x] No new code climate issues added

### Logging
- [x] Are logs being produced at appropriate log levels (debug, info, warn, error, fatal)?

### Error Handling
- [x] Are errors being caught and handled gracefully?

### CODE REVIEW

File: app/services/events/decision_review_create/update_vacols_on_optin.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module Events::DecisionReviewCreate::UpdateVacolsOnOptin
+  # Updates the Caseflow and VACOLS DB when Legacy Issues Optin to AMA
+  # the decision review argument being passed in can either be a Higher Level Review or a Supplemental Claim
+  # the decision review hash must have a benefit type.
+  def self.perform!(decision_review)
```
Comment (nkutub) [inline app/services/events/decision_review_create/update_vacols_on_optin.rb:7]: I would use run! vs preform! , preform! is usually reserved for jobs

File: app/services/events/decision_review_create/update_vacols_on_optin.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    if decision_review.legacy_opt_in_approved
+      LegacyOptinManager.new(decision_review: decision_review).process!
+    end
+  # catch the error that is produced
+  rescue StandardError => error
+    raise error
```
Comment (nkutub) [inline app/services/events/decision_review_create/update_vacols_on_optin.rb:13]: lets make a new DecisionReviewCreateVacolsOnOptinError here and raise that

File: spec/services/events/decision_review_create/update_vacols_on_optin_spec.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      RequestStore.store[:current_user] = user
+    end
+    it "Does not run perform if SOC/SSOC is not approved" do
+      subject
+      expect(subject).to be_nil
+    end
+    it "Updates VACOLS ISSUE Dispostition Code" do
+      higher_level_review.update!(legacy_opt_in_approved: true)
+      subject
+      expect(vacols_issue("DRCTEST", 1).issdc).to eq(LegacyIssueOptin::VACOLS_DISPOSITION_CODE)
```
Comment (nkutub) [inline spec/services/events/decision_review_create/update_vacols_on_optin_spec.rb:79]: this is great spec work but you are going beyond what should be testing here , we should only test the outcome of this service class of that the fact the LegacyOptinManager.new(decision_review: decision_review).process! was called , we don't need to verify it did its job. We end up making a very dependent spec on how LegacyOptinManager works so if it changes how it records in VACOLS then this spec will break. This spec should not be concerned on how LegacyOptinManager changes VACOLS only that it was called and trusted to do its job. we can simplify the spec by stubbing this call using double

File: app/services/events/decision_review_create/update_vacols_on_optin.rb (hunk 1 above)
Comment (Jruuuu) [reply app/services/events/decision_review_create/update_vacols_on_optin.rb:7]: got it! working it now!

File: app/services/events/decision_review_create/update_vacols_on_optin.rb (hunk 2 above)
Comment (nkutub) [reply app/services/events/decision_review_create/update_vacols_on_optin.rb:13]: put  error def in the Error.rb

Comment (Jruuuu) [reply app/services/events/decision_review_create/update_vacols_on_optin.rb:13]: got it working it now

File: spec/services/events/decision_review_create/update_vacols_on_optin_spec.rb (hunk 3 above)
Comment (Jruuuu) [reply spec/services/events/decision_review_create/update_vacols_on_optin_spec.rb]: Got it! Thank you for your help!

**Review by nkutub — APPROVED** [review]

**Pull request merged**